src/index/hnsw.cpp

```diff
@@ -209,6 +213,8 @@ HNSW::knn_search(const DatasetPtr& query,
         CHECK_ARGUMENT(k > 0, fmt::format("k({}) must be greater than 0", k))
         k = std::min(k, GetNumElements());
 
+        std::shared_lock lock(rw_mutex_);
+
         // check search parameters
         auto params = HnswSearchParameters::FromJson(parameters);
         alg_hnsw->setEf(std::max(params.ef_search, k));
```

Comment:
In 'shared_lock' scope here  have thread-safe issue, move parameter 'ef' to stack variable ?